### PR TITLE
Profalux covers: switch to using cluster 0x102, add MAI-ZTP20C remote

### PR DESCRIPTION
**Pull request by sven337**
This change references remote model MAI-ZTP20C.
More importantly, it switches the control of covers to using the dedicated ClosuresWindowCovering (0x102) cluster. Sniffing Zigbee traffic of my remotes show that the remotes use this cluster to control the covers, and not 0x8 as currently implemented. The compliance document for the Profalux motors at https://csa-iot.org/csa_product/ac-motor-for-window-coverings/ also lists that cluster and does NOT list 0x6/0x8.

So I think the correct way to control these covers is 0x102, and certainly it's the way that seems to work with my shutters (all bought in 2023).

This change is likely to break existing setups if older revisions of these motors did not expose 0x102.

0x102 is on endpoint 2, while 0x6/0x8 are on endpoint 1. An ideal implementation would use 0x102 when it exists, and 0x8 the rest of the time, but I don't know how to implement that.

This change needs testing by other Profalux users!
Looking into https://github.com/Koenkk/zigbee2mqtt/issues/17674 is what prompted me to sniff traffic to figure out what was wrong.

### CODE REVIEW

**Comment by sven337** [comment]
@ralmn @rclsilver  any chance one of you could take a look? Do your remotes also use 0x102?
Thanks

**Comment by ralmn** [comment]
I haven't had time to test it yet. I'll test this on Friday or this weekend. 

On the other hand I had read in a technical docs of Profalux that it recommended to use the control via brightness...  I will try to find it

**Comment by sven337** [comment]
Not stale - this is still the best thing to do vs. existing, as far as I can tell.

**Comment by rclsilver** [comment]
I'll try to test it this week

**Comment by sven337** [comment]
@Koenkk, do you have any suggestion on 
>0x102 is on endpoint 2, while 0x6/0x8 are on endpoint 1. An ideal implementation would use 0x102 when it exists, and 0x8 the rest of the time, but I don't know how to implement that. 
?

In case the older versions of this hardware only expose the older clusters -- I have no ability to check but suspect it /might/ be the case.

**Comment by sven337** [comment]
We don't know which models may or may not support 0x102. I suspect that all of them do (per the zigbee conformance document I linked to in my first message). So I can't split by model, I would need to split based on supported clusters, but I found no pre-existing example of that in Zigbee2MQTT.

At any rate, I still think that my change is correct, but it carries a risk of regressing on older hardware that I could not test. It wouldn't be the end of the world to fix it if it happened.

**Comment by rclsilver** [comment]
OK, I'll try to deploy your version once my baby is woke up

**Comment by Koenkk** [comment]
Can you provide the data/database.db version of both the old an new version of this device?

**Comment by sven337** [comment]
I don't have access to the "old" version of the device, sorry. I have no proof that older versions of the device don't support cluster 0x102. It's a "just in case" situation.

**Comment by Koenkk** [comment]
If it works for @rclsilver this is OK (since he added support for this device in https://github.com/Koenkk/zigbee-herdsman-converters/pull/3855)

**Comment by rclsilver** [comment]
Hi, @Koenkk you will find my [database.db.zip](https://github.com/Koenkk/zigbee-herdsman-converters/files/11886803/database.db.zip) here.

I can't test at the moment because my work and my baby, but if I'm able to quickly build a Docker image, I would be able to try your fix. @sven337 maybe could you prepare a Dockerfile for me?

**Comment by sven337** [comment]
I don't use Docker. But it should just be a matter of replacing the profalux.js file for a quick test.

**Comment by rclsilver** [comment]
Well, I wrote this Dockerfile inside your branch:
```
FROM koenkk/zigbee2mqtt:1.31.2

RUN rm -rf /app/node_modules/zigbee-herdsman-converters

COPY . /app/node_modules/zigbee-herdsman-converters
```

This provides a zigbee2mqtt 1.31.2 docker image with your code.

I'll deploy it on my production tomorrow (restart of z2m make all my covers moving).

Stay tunned

**Comment by Koenkk** [comment]
@sven337 can you provide the data/database.db of the new version?

@rclsilver it is easier to use this ext converter: https://gist.github.com/Koenkk/ffae327b02913d375f87014e0827e772

Could you check if the issue is fixed with the following external converter: 

- save this as file next to `configuration.yaml` as `ext_converter.js`
- add it to `configuration.yaml`:
```yaml
external_converters:
  - ext_converter.js
```
- start z2m, check in the frontend if the device description is "Cover CUSTOM" now
- check if device still works.

**Comment by sven337** [comment]
Here's the database entry for the new version of the HW:

`{"id":13,"type":"Router","ieeeAddr":"0x20918a0000188731","nwkAddr":32064,"manufId":4368,"manufName":"Profalux","powerSource":"Mains (single phase)","modelId":"MOT-C1Z06F\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000","epList":[1,2,242],"endpoints":{"1":{"profId":260,"epId":1,"devId":512,"inClusterList":[0,3,4,5,6,8,10,256,64533,64544,64545],"outClusterList":[10,64544],"clusters":{"genBasic":{"attributes":{"modelId":"MOT-C1Z06F\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000","manufacturerName":"Profalux","powerSource":1,"zclVersion":8,"appVersion":34,"hwVersion":29,"dateCode":"20000101","swBuildId":"Release"}},"genLevelCtrl":{"attributes":{"currentLevel":1}}},"binds":[{"cluster":8,"type":"endpoint","deviceIeeeAddress":"0xe0798dfffef8bab4","endpointID":1}],"configuredReportings":[{"cluster":8,"attrId":0,"minRepIntval":0,"maxRepIntval":3600,"repChange":1}],"meta":{}},"2":{"profId":260,"epId":2,"devId":514,"inClusterList":[0,1,2,3,4,5,9,10,258,2821,64514,64533,64544,64545],"outClusterList":[10,25,32,64544],"clusters":{"genBasic":{"attributes":{"zclVersion":8}},"closuresWindowCovering":{"attributes":{"currentPositionLiftPercentage":100,"currentPositionTiltPercentage":255}}},"binds":[{"cluster":258,"type":"endpoint","deviceIeeeAddress":"0xe0798dfffef8bab4","endpointID":1}],"configuredReportings":[{"cluster":258,"attrId":8,"minRepIntval":1,"maxRepIntval":62000,"repChange":1}],"meta":{}},"242":{"profId":41440,"epId":242,"devId":102,"inClusterList":[33],"outClusterList":[33],"clusters":{},"binds":[],"configuredReportings":[],"meta":{}}},"appVersion":34,"hwVersion":29,"dateCode":"20000101","swBuildId":"Release","zclVersion":8,"interviewCompleted":true,"meta":{"configured":-1514179447},"lastSeen":1687977782265,"defaultSendRequestWhen":"immediate"}`

**Comment by rclsilver** [comment]
I deployed the external converters as described and I have the "Cover CUSTOM" description.

Sadly, I'm not able to use my covers anymore:

```
Zigbee2MQTT:error 2023-06-29 21:18:15: Failed to configure 'salon_cover', attempt 2 (TypeError: Cannot read property 'bind' of undefined
    at Object.bind (/app/node_modules/zigbee-herdsman-converters/lib/reporting.js:34:24)
    at Object.configure (/app/dist/util/externally-loaded.js:26:25)
    at Configure.configure (/app/lib/extension/configure.ts:115:37)
    at EventEmitter.<anonymous> (/app/lib/extension/configure.ts:80:62)
    at EventEmitter.emit (events.js:412:35)
    at EventBus.emitLastSeenChanged (/app/lib/eventBus.ts:52:22)
    at Controller.<anonymous> (/app/lib/zigbee.ts:66:27)
    at Controller.emit (events.js:400:28)
    at Controller.selfAndDeviceEmit (/app/node_modules/zigbee-herdsman/src/controller/controller.ts:483:14)
    at Controller.onZclOrRawData (/app/node_modules/zigbee-herdsman/src/controller/controller.ts:694:18))
```

And:
```
Device 'bureau_cover' has no endpoint 'undefined'
```

Do not hesitate if you want me more tests

**Comment by sven337** [comment]
Can you show me the database entry for your cover, or maybe a screenshot of the "clusters" tab?

**Comment by rclsilver** [comment]
Yes sure, I sent my db here: https://github.com/Koenkk/zigbee-herdsman-converters/pull/5788#issuecomment-1610137335

**Comment by sven337** [comment]
Thanks, here's your entry:
`{"id":17,"type":"Router","ieeeAddr":"0x20918a00000dacd6","nwkAddr":64618,"manufId":4368,"powerSource":"Mains (single phase)","epList":[1],"endpoints":{"1":{"profId":260,"epId":1,"devId":512,"inClusterList":[0,3,4,5,6,8,10,21,256,64544,64545],"outClusterList":[3,64544],"clusters":{"genBasic":{"attributes":{"powerSource":1,"zclVersion":1,"dateCode":"20180723102c104"}},"genLevelCtrl":{"attributes":{"currentLevel":254}}},"binds":[{"cluster":8,"type":"endpoint","deviceIeeeAddress":"0x00158d0006a3fd34","endpointID":1}],"configuredReportings":[{"cluster":8,"attrId":0,"minRepIntval":0,"maxRepIntval":3600,"repChange":1}],"meta":{}}},"dateCode":"20180723102c104","zclVersion":1,"interviewCompleted":true,"meta":{"configured":734529022},"lastSeen":1687892417674,"defaultSendRequestWhen":"immediate"}
`

No endpoint 2, no cluster 0x102. This confirms that older versions of the HW need to use cluster 0x8 on endpoint 1.
@Koenkk can you suggest a way to use 0x102 on new HW, 0x8 on old HW, based on the database entries here?
Thanks

**Comment by Koenkk** [comment]
@sven337 instead of using zigbeeModel to identify, use `fingerprint` and create 2 separate definitions for this device. Example: https://github.com/Koenkk/zigbee-herdsman-converters/blob/1ff7011ca707151e5f752769d38a03244dc62fba/src/devices/ecodim.ts#L12

**Comment by rclsilver** [comment]
@sven337 if you want me to test, do not hesitate to tell me

**Comment by sven337** [comment]
@Koenkk I'd like your input on how to solve this.
toZigbee.js's thermostat_system_mode reads from constants.thermostatSystemModes
It feels wasteful to duplicate the code and make a acova_thermostat_system_mode that reads from a new constants.acovaThermostatSystemModes
What do you think of using the "meta" parameter to store the system mode array to be used? I see that tuya.js does this
`        meta: {tuyaThermostatSystemMode: tuya.thermostatSystemModes2, tuyaThermostatPreset: tuya.thermostatPresets},`

Do you agree with that plan or would you rather I did it a different way?

Thanks

**Comment by mauduich** [comment]
Hello,

I have MOT-C1Z10C, so I created an external converter to test it. It works fine for me.
The only feature missing is the get state. I don't know how to add it or even if it's possible.

I'm in a weird situation where two of my profalux covers, after a one minute reset, did not work with the previous version. This one allow me to open/close/stop my two different types cover: MOT-C1Z10C & MOT-C1Z06F.

If I can help further in anyway, just tell me.

**Comment by mauduich** [comment]
and fo my remotes, I have MAI-ZTM20C. It works fine too.

**Comment by sven337** [comment]
I haven't got a chance to implement this "cleanly" following https://github.com/Koenkk/zigbee-herdsman-converters/pull/5788#issuecomment-1614919010 yet.
If you'd like to try...

Glad to hear this helps you.

**Comment by sven337** [comment]
OK, I am very confused with the .js -> .ts change. @Koenkk could you please help me out?
I've updated my change by editing zigbee2mqtt/node_modules/zigbee-herdsman-converters/devices/profalux.js directly (and @rclsilver I'll need some new testing from you if you're OK), but zigbee-herdsman-converters seems like it uses .ts and not .js and these are not strictly equivalent.
Koen if you wouldn't mind, would you please take a look at my PR and see what I did wrong? 
Thanks

File: src/devices/profalux.ts [hunk 1]
```diff
@@ -18,14 +18,35 @@ const definitions: Definition[] = [
...
+    },
+    {
+        // Identify older covers based on their fingerprint. These do not
+        // expose closuresWindowCovering and need to use genLevelCtrl
+        // instead. Sniffing a remote would be welcome to confirm that this
+        // is the right thing to do.
         fingerprint: [{manufacturerID: 4368, endpoints: [{ID: 1, profileID: 260, deviceID: 512,
             inputClusters: [0, 3, 4, 5, 6, 8, 10, 21, 256, 64544, 64545], outputClusters: [3, 64544]}]}],
-        model: 'NSAV061',
+        model: 'NSAV061 (old)',
```
Comment (Koenkk) [inline src/devices/profalux.ts:49]: Please use `NSAV061_old` here

Comment (sven337) [reply src/devices/profalux.ts]: Done

**Comment by sven337** [comment]
> @sven337 for the changes in this pr the syntax between .ts and .js should be the same. Editing `devices/profalux.js` directly for testing is fine.

OK, this kinda works, but it's not the same. 
.ts:
    `    toZigbee: [tz.cover_state, tz.cover_position_tilt],`
.js:
`        toZigbee: [toZigbee_1.default.cover_state, toZigbee_1.default.cover_position_tilt],`
I don't know if that's expected.

Anyway, what you see in the PR now is tested for the new version. The old version should not behave any differently but it would be nice to get confirmation from @rclsilver or somebody else.

**Comment by rclsilver** [comment]
Hi @sven337, I'll check as soon as possible.

I see you set "my covers" to `_old`, maybe you could set _v1 and _v2 instead?

**Comment by sven337** [comment]
Your covers are the old version, so I think it's correct.
When were yours purchased?

**Comment by rclsilver** [comment]
Your covers will be "old" one day, mine will be _old_old and yours _old? lol

**Comment by sven337** [comment]
The covers you have do not match Profalux's own Zigbee compliance documents since they use a different cluster, so they must be some sort of "preliminary" firmware.
I don't have a contact at Profalux so it's all guesswork, but it seems that they released a first version of their Zigbee motors which did not advertise a Zigbee model and used generic clusters (that's what you have), and then released an updated version that properly advertises a Zigbee model and uses the appropriate cluster for covers, and matches their ZCA compliance docs. So the updated version conforms to what the Zigbee standard expects of covers and it's unlikely there will be breaking changes in the future for that reason; whereas the version you have is very special. 

Chronologically the version you have is older, and arguably less conformant, so I don't think "old" is inappropriate at all.
Note that we don't know the real model number on yours, but I bet it's going to be something other than MOT-C1Zxx{C,F}. Maybe -C1Z06A? (the figures are the torque and I think the last letter is the version, but I am not certain)

Anyway, would you please test the PR and let me know if it works for you? I don't want to break your covers but I can't test them myself.

**Comment by sven337** [comment]
> I have MOT-C1Z10C, so I created an external converter to test it. It works fine for me. The only feature missing is the get state. I don't know how to add it or even if it's possible.

It's done in the latest version of the PR, care to verify?

**Comment by sven337** [comment]
Let's go with _v1 if that suits you. 
Note that this is just a name. NSAV061 isn't correct anyway: it means "notice SAV numéro 061", it's not a model number. The correct thing to do /if we had enough information/ would be to call yours MOT-C1ZxxA, and the newer ones MOT-C1ZxxC, MOT-C1ZxxF. Thing is, I don't know the model number of your covers.
The newer ones I'm certain of. Yours, not so much. 

So - I've pushed an update to use correct model numbers. @Koenkk, is this model string:
**MOT-C1ZxxC/F** acceptable to you? I am not certain if you have syntax rules there

@rclsilver please test and report back to confirm that this still works for you.

**Comment by Koenkk** [comment]
@sven337 `MOT-C1ZxxC/F` doesn't yield any results when googling on it, so I propose to revert the the old one.

**Comment by sven337** [comment]
I was more asking if it was a problem syntax wise :)

The real model numbers are:
MOT-C1Z06C
MOT-C1Z10C
MOT-C1Z20C
MOT-C1Z30C
MOT-C1Z06F
MOT-C1Z10C
MOT-C1Z20C
MOT-C1Z30C

I tried to use something that generically represented these names. But we shouldn't use the old one because it's not a model number (it's the name of a PDF document not of the motor).

**Comment by Koenkk** [comment]
I see, then let's leave it to `MOT-C1ZxxC/F`. Let me know once this is ready for merge (I left one more comment but this rest looks good)

**Comment by sven337** [comment]
@rclsilver has not tested it yet, maybe @mauduich can test and report back? (since you tested the first time, the state reporting should be working now)

**Comment by rclsilver** [comment]
Yes, I haven't tested yet because the content I tried to set in ext_converter didn't worked and I hadn't any time to check why it did'nt work yet

**Comment by sven337** [comment]
> Yes, I haven't tested yet because the content I tried to set in ext_converter didn't worked and I hadn't any time to check why it did'nt work yet

Probably the .js/.ts thing that has given me trouble. Here's a .js to replace yours (remove the txt extension)
[profalux.js.txt](https://github.com/Koenkk/zigbee-herdsman-converters/files/12526049/profalux.js.txt)

**Comment by mauduich** [comment]
> @rclsilver has not tested it yet, maybe @mauduich can test and report back? (since you tested the first time, the state reporting should be working now)

Hello,

On my side everything is working fine with the following external converters.
[MOT-C1Z10C.js.txt](https://github.com/Koenkk/zigbee-herdsman-converters/files/12528253/MOT-C1Z10C.js.txt)
[MAI-ZTM20C.js.txt](https://github.com/Koenkk/zigbee-herdsman-converters/files/12528254/MAI-ZTM20C.js.txt)

@sven337 it's not exactly what you develop, but it is based on your work.
If you want me to test your external converter, I could take a look by the end of the week.

Regards

**Comment by sven337** [comment]
This is a pull request: I intend my work to become part of zigbee-herdsman-converters
So it would be useful for you to test it, since it might become what you'll use :)
Looking at your own external converters, I believe everything will work as we're doing the same thing.
Thanks.

**Comment by sven337** [comment]
@Koenkk this is slowly bit-rotting because I've had trouble enrolling external testers, but I think it's ready to go; it's been tested to the best of my abilities!

**Comment by Koenkk** [comment]
thanks @sven337 , this pr will be included in todays release.

**Comment by mauduich** [comment]
> This is a pull request: I intend my work to become part of zigbee-herdsman-converters So it would be useful for you to test it, since it might become what you'll use :) Looking at your own external converters, I believe everything will work as we're doing the same thing. Thanks.

@sven337 I confirm that everything is working fine on my side. I updated to z2m last version few days ago.
Just my remote that is not recognize as a supported device, but it's not a big matter.
Do you want me to add this support or I let you lead it based on the external converter published on a comment on early September ?

**Comment by sven337** [comment]
My apologies!  I did not pay attention that you had a ZTM20C. What is this remote? Do you have a picture?
I have ZTP20.
You can add support to ZTM20C if you like (you can test it directly, it's better!). If it gives you trouble I can do it.

**Comment by mauduich** [comment]
It's this one https://ateliervolet.fr/cdn/shop/products/emetteur-radio-profalux-zigbee-mural-1-canal_2048x2048.png?v=1582253062

I will do it, from what I understand I only have to add  Zigbee model here:
Capture d’écran 2023-10-20 à 13 37 26

Or do you think I should create brand new object for this remote ? To ensure picture match the remote.
I think P is for portable and M for mural.

**Comment by sven337** [comment]
Given that the remote is useless in Z2M I don't think it's useful to make a separate device. Sure the picture won't match but it should not matter, and I guess the documentation can be updated to say that both wall-mounted and portable remotes are covered?

**Comment by elriando** [comment]
Hello, I also have the ZTM20C remote control. Can you modify it, make a request for modification or tell me how I can also manage to modify the P into M, thank you!

**Comment by sven337** [comment]
See comment above from @mauduich, what he's showing in the picture is what you need.
But this serves no purpose other than showing the right picture.

**Comment by elriando** [comment]
Thanks for the super quick response. I would like to simply add “MAI-ZTM20C” but I don’t know the file where mauduich finds it.

**Comment by mauduich** [comment]
Hello,

I have open a pull request to add MAI-ZTM20C remote #6835

**Pull request merged**